src/event.rs

```diff
@@ -0,0 +1,616 @@
+					.add_event(Event::PaymentSuccessful { payment_hash })
+					.expect("Failed to push to event queue");
+			}
+			LdkEvent::PaymentFailed { payment_hash, .. } => {
+				log_info!(
+					self.logger,
+					"Failed to send payment to payment hash {:?}: exhausted payment retry attempts",
+					hex_utils::to_string(&payment_hash.0)
+				);
+
+				let mut payments = self.outbound_payments.lock().unwrap();
+				if payments.contains_key(&payment_hash) {
+					let payment = payments.get_mut(&payment_hash).unwrap();
+					assert_eq!(payment.status, PaymentStatus::Pending);
+					payment.status = PaymentStatus::Failed;
+				}
+				self.event_queue
+					.add_event(Event::PaymentFailed { payment_hash })
+					.expect("Failed to push to event queue");
+			}
+
+			LdkEvent::PaymentPathSuccessful { .. } => {}
+			LdkEvent::PaymentPathFailed { .. } => {}
+			LdkEvent::ProbeSuccessful { .. } => {}
+			LdkEvent::ProbeFailed { .. } => {}
+			LdkEvent::HTLCHandlingFailed { .. } => {}
+			LdkEvent::PendingHTLCsForwardable { time_forwardable } => {
+				let locked_runtime = self.tokio_runtime.read().unwrap();
+				if locked_runtime.as_ref().is_none() {
+					return;
+				}
+
+				let forwarding_channel_manager = self.channel_manager.clone();
+				let min = time_forwardable.as_millis() as u64;
+
+				locked_runtime.as_ref().unwrap().spawn(async move {
+					let millis_to_sleep = thread_rng().gen_range(min..min * 5) as u64;
+					tokio::time::sleep(Duration::from_millis(millis_to_sleep)).await;
+
+					forwarding_channel_manager.process_pending_htlc_forwards();
+				});
+			}
+			LdkEvent::SpendableOutputs { outputs } => {
+				// TODO: We should eventually remember the outputs and supply them to the wallet's coin selection, once BDK allows us to do so.
+				let destination_address = self.wallet.get_new_address().unwrap();
+				let output_descriptors = &outputs.iter().map(|a| a).collect::<Vec<_>>();
+				let tx_feerate =
+					self.wallet.get_est_sat_per_1000_weight(ConfirmationTarget::Normal);
+				let spending_tx = self
+					.keys_manager
+					.spend_spendable_outputs(
+						output_descriptors,
+						Vec::new(),
+						destination_address.script_pubkey(),
+						tx_feerate,
+						&Secp256k1::new(),
+					)
+					.unwrap();
+				self.wallet.broadcast_transaction(&spending_tx);
+			}
+			LdkEvent::OpenChannelRequest { .. } => {}
+			LdkEvent::PaymentForwarded {
+				prev_channel_id,
+				next_channel_id,
+				fee_earned_msat,
+				claim_from_onchain_tx,
+			} => {
+				let read_only_network_graph = self.network_graph.read_only();
+				let nodes = read_only_network_graph.nodes();
+				let channels = self.channel_manager.list_channels();
+
+				let node_str = |channel_id: &Option<[u8; 32]>| match channel_id {
+					None => String::new(),
+					Some(channel_id) => match channels.iter().find(|c| c.channel_id == *channel_id)
+					{
+						None => String::new(),
+						Some(channel) => {
+							match nodes.get(&NodeId::from_pubkey(&channel.counterparty.node_id)) {
```

Comment:
May be able to reduce indentation by merging these lines, assuming it fits in line length constraint.

Reply:
Done.